src/tir/transforms/vectorize_loop.cc

```diff
@@ -72,6 +72,126 @@ inline PrimExpr BroadcastTo(PrimExpr e, int lanes, bool is_scalable) {
   return Broadcast(e, CreateNewLanes(is_scalable, lanes));
 }
 
+bool EnableBufferLevelPredication() {
```

Comment:
I like the fallback, where it checks whether a target supports SVE in order to apply the vectorization if not specified in the pass context.  However, this function should accept a `Target` argument from the calling scope rather than using `Target::Current()`, because the `Target::Current()` may be overridden. 

The override can be either at the function level (`prim_func->GetAttr<Target>(tvm::attr::kTarget)`), or for individual blocks of TIR (an `AttrStmt` with `attr_stmt->attr_key==tvm::attr::kTarget`).

Reply:
This will require updating `arith::TargetHasSVE()` as well.  Apologies for not catching it earlier in #16893.

Reply:
Thanks, that makes sense. Do you know if there is any general infrastructure for keep track of the current target (which takes into account this override functionality) from within a pass? Otherwise I feel we will be duplicating this functionality in multiple places. I was thinking something similar to: `LexicalOnDeviceMixin` (assuming I understood it correctly)

Reply:
To my knowledge, there isn't a general utility that would provide it, though I definitely agree on its usefulness.

There's currently a couple similar utilities, such as [`IRMutatorWithAnalyzer`](https://github.com/apache/tvm/blob/main/src/arith/ir_mutator_with_analyzer.h#L45) and [`IRVisitorWithAnalyzer`](https://github.com/apache/tvm/blob/main/src/arith/ir_visitor_with_analyzer.h#L35), which allow context-dependent algebraic simplifications. The easiest way to implement such a utility for target-aware simplifications would be to extract out the context-dependent target tracking from an existing visitor/mutator (e.g. [here](https://github.com/apache/tvm/blob/main/src/tir/transforms/lower_device_kernel_launch.cc#L65), used in the `LowerDeviceKernelLaunch` transform) into a `IRVisitorWithTarget` and `IRMutatorWithTarget`, which is then subclassed again by each transform that requires it.

For long-term maintainability, I like your suggestion of mixins much more than the subclasses, because they could be composed together much more easily.  Implemented with subclasses, a hypothetical transform that requires both the context-dependent target and context-dependent algebraic simplifications would have incorrect behavior from multiple inheritance, and would need to re-implement one or the other functionalities.

Unfortunately, I don't think the current `tir::StmtVisitor` API could be used to support mixins, either through multiple inheritance or through composition.  The main issue is that the implementable virtual functions (e.g. `tir::StmtVisitor::VisitStmt_(const tir::ForNode*)`) aren't granular enough.  In order to compose separate transforms that provide context-dependent information, it would require separate virtual functions for pre-visit and post-visit.  I've toyed around with different variations of the TIR stmt/expr visitor/mutator implementations, but haven't had the time to make anything PR-able.

TL;DR: It doesn't currently exist, would be very useful to have, and a subclass would be sufficient for this PR, though long-term maintainability concerns may appear at some point in the future.

Reply:
Thanks for the detailed description. To save this PR exploding further, I've added the minimum changes necessary for the vectorizer pass. Perhaps we can clean this up and fix the target checks in the analyzer in a separate PR?